test/module/irohad/multi_sig_transactions/mst_processor_test.cpp

```diff
@@ -0,0 +1,262 @@
+#include "logger/logger.hpp"
+#include "module/irohad/multi_sig_transactions/mst_mocks.hpp"
+#include "module/irohad/multi_sig_transactions/mst_test_helpers.hpp"
+#include "multi_sig_transactions/mst_processor_impl.hpp"
+#include "multi_sig_transactions/storage/mst_storage_impl.hpp"
+
+auto log_ = logger::log("MstProcessorTest");
+
+using namespace std;
+using namespace iroha;
+using namespace iroha::model;
+using namespace framework::test_subscriber;
+
+using testing::Return;
+using testing::_;
+
+class TestCompleter : public Completer {
+  bool operator()(const DataType transaction) const override {
+    return transaction->signatures.size() >= transaction->quorum;
+  }
+
+  bool operator()(const DataType &tx, const TimeType &time) const override {
+    return tx->created_ts < time;
+  }
+};
+
+class MstProcessorTest : public testing::Test {
+ public:
+  // --------------------------------| fields |---------------------------------
+
+  Peer own_peer = makePeer("iam", "1");
+  /// propagation subject, useful for propagation control
+  rxcpp::subjects::subject<PropagationStrategy::PropagationData>
+      propagation_subject;
+  /// use effective implementation of storage
+  shared_ptr<MstStorage> storage;
+  shared_ptr<FairMstProcessor> mst_processor;
+
+  // ---------------------------------| mocks |---------------------------------
+
+  shared_ptr<MockMstTransport> transport;
+  shared_ptr<MockPropagationStrategy> propagation_strategy;
+  shared_ptr<MockTimeProvider> time_provider;
+
+ protected:
+  void SetUp() override {
+    transport = make_shared<MockMstTransport>();
+    storage = make_shared<MstStorageStateImpl>(own_peer,
+                                               make_shared<TestCompleter>());
+
+    propagation_strategy = make_shared<MockPropagationStrategy>();
+    EXPECT_CALL(*propagation_strategy, emitter())
+        .WillOnce(Return(propagation_subject.get_observable()));
+
+    time_provider = make_shared<MockTimeProvider>();
+    EXPECT_CALL(*time_provider, getCurrentTime()).WillRepeatedly(Return(111));
+
+    mst_processor = make_shared<FairMstProcessor>(
+        transport, storage, propagation_strategy, time_provider);
+  }
+};
+
+/**
+ * @given initialised mst processor
+ * AND wrappers on mst observables
+ *
+ * @when insert not-completed transaction
+ *
+ * @then check that:
+ * state not updated
+ * AND absent prepared transactions
+ * AND absent expired transactions
+ */
+TEST_F(MstProcessorTest, notCompletedTransactionUsecase) {
+  // ---------------------------------| given |---------------------------------
+  auto onStateUpdateWrapper =
+      make_test_subscriber<CallExact>(mst_processor->onStateUpdate(), 0);
+  onStateUpdateWrapper.subscribe();
+
+  auto onPreparedTransactionsWrapper = make_test_subscriber<CallExact>(
+      mst_processor->onPreparedTransactions(), 0);
+  onPreparedTransactionsWrapper.subscribe();
+
+  auto onExpiredTransactionsWrapper = make_test_subscriber<CallExact>(
```

Comment:
just an advice: this triplet of make-subscibe calls can be extracted to function that return `std::tuple<.., .., ..>`
then code will be much easier to read

Reply:
`make_test_subscriber` return only one thing. Could you explain more details about your issue?

Reply:
```
std::tuple<T1, T2, T3> f(int a, int b, int c) {
  std::tuple t{make_test_subscriber<CallExact>(mst_processor->onStateUpdate(), a),
    make_test_subscriber<CallExact>(mst_processor->onPreparedTransactions(), b),
    make_test_subscriber<CallExact>(mst_processor->onExpiredTransactions(), c)};
  std::get<0>(t).subscribe();
  std::get<1>(t).subscribe();
  std::get<2>(t).subscribe();
}
```
something like this